### PR TITLE
Unique selfregister shortname

### PR DESCRIPTION
**Pull request by vmcj**
This is a possible fix for: https://github.com/DOMjudge/domjudge/issues/841

It still needs some styling cleanup but I would like some feedback if we indeed want this.

Also we don't have password restrictions for this form, which might be a problem for how other people use this where the people self registering are not spectators (this is how I use them for my prelims) but "real" contestants. But this is not relevant for this PR.

### CODE REVIEW

**Comment by nickygerritsen** [comment]
For `Eindhoven University of Technology` this would become `Technology0` right? (technology is longer than eindhoven).
That doesn't sound like the correct one. Also, why always append a number?

**Comment by vmcj** [comment]
> For `Eindhoven University of Technology` this would become `Technology0` right? (technology is longer than eindhoven). That doesn't sound like the correct one. Also, why always append a number?

To make sure that we get a unique result, as the heuristic is not that strong (you prove that point). As the shortname is only internal I permitted myself a bit more freedom in this generation.

And although at first I agreed that your case should have "Eindhoven0" (or even "Eindhoven") but in case of:

Vrije Universiteit Amsterdam
Universiteit van Amsterdam
Hoge school van Amsterdam
High school van Amsterdam
Something in Amsterdam
You get the point it's in Amsterdam

I would say we would not prefer to get the city. 

So maybe combine the longest and shorter element? I would say that would almost always handle the case we want?

**Comment by nickygerritsen** [comment]
> > For `Eindhoven University of Technology` this would become `Technology0` right? (technology is longer than eindhoven). That doesn't sound like the correct one. Also, why always append a number?
> 
> To make sure that we get a unique result, as the heuristic is not that strong (you prove that point). As the shortname is only internal I permitted myself a bit more freedom in this generation.
> 
> And although at first I agreed that your case should have "Eindhoven0" (or even "Eindhoven") but in case of:
> 
> Vrije Universiteit Amsterdam Universiteit van Amsterdam Hoge school van Amsterdam High school van Amsterdam Something in Amsterdam You get the point it's in Amsterdam
> 
> I would say we would not prefer to get the city.
> 
> So maybe combine the longest and shorter element? I would say that would almost always handle the case we want?

If the shortname is only used internally, can't we set a UUID in it upon registration? 🤔

**Comment by vmcj** [comment]
> > > For `Eindhoven University of Technology` this would become `Technology0` right? (technology is longer than eindhoven). That doesn't sound like the correct one. Also, why always append a number?
> > 
> > 
> > To make sure that we get a unique result, as the heuristic is not that strong (you prove that point). As the shortname is only internal I permitted myself a bit more freedom in this generation.
> > And although at first I agreed that your case should have "Eindhoven0" (or even "Eindhoven") but in case of:
> > Vrije Universiteit Amsterdam Universiteit van Amsterdam Hoge school van Amsterdam High school van Amsterdam Something in Amsterdam You get the point it's in Amsterdam
> > I would say we would not prefer to get the city.
> > So maybe combine the longest and shorter element? I would say that would almost always handle the case we want?
> 
> If the shortname is only used internally, can't we set a UUID in it upon registration? 🤔

I still prefer to have something which one could remember if you care about the self-registered contestants for another reason.

**Comment by thijskh** [comment]
I prefer [this solution](https://github.com/DOMjudge/domjudge/issues/841#issuecomment-650563633) over 'magic' that may some of the time, may not other times, generate a meaningful shortname.

Usually institutions already have such a commonly used short version of their official name, e.g. Technische Universiteit Eindhoven = "TUE" or "TU/e" (sic).

The suggested AI fails on Hogeschool Utrecht and also fails on Universiteit Utrecht if Hogeschool Utrecht also participates.

File: webapp/src/Form/Type/UserRegistrationType.php [hunk 1]
```diff
@@ -139,6 +139,15 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
...
                     'mapped' => false,
+                ])
+                ->add('affiliationShortName', TextType::class, [
+                    'label' => false,
+                    'required' => false,
+                    'attr' => [
+                        'placeholder' => 'Affiliation shortname',
+                        'maxlength' => '32',
+                    ],
+                    'mapped' => false,
```
Comment (nickygerritsen) [inline webapp/src/Form/Type/UserRegistrationType.php:150]: Useless spaces after the line

**Review by nickygerritsen — APPROVED** [review]

File: webapp/src/Controller/SecurityController.php [hunk 2]
```diff
@@ -145,10 +145,11 @@ public function registerAction(
             if ($this->config->get('show_affiliations')) {
                 switch ($registration_form->get('affiliation')->getData()) {
                     case 'new':
+                        $shortNameLength = 32;
```
Comment (thijskh) [inline webapp/src/Controller/SecurityController.php:148]: This variable is introduced but unused.

Comment (thijskh) [reply webapp/src/Controller/SecurityController.php:148]: FYI, the shortname is already truncated with a hard `substr(,0,32)` in the Entity/TeamAffiliation class. So the form validation plus that should ensure it's never too long.

**Review by thijskh — APPROVED** [review]
Good work, great to see all those tests!

**Pull request merged**